### PR TITLE
shell: shell_uart: add ring_buffers and interrupt support

### PR DESCRIPTION
**Pull request by nordic-krch**
Improved RX path to use ring buffer for incoming data instead of single
byte buffer. Improved TX path to use ring buffer. Added support for
asynchronous UART API (interrupts).

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

### CODE REVIEW

**Comment by nordic-krch** [comment]
It is rebased on #10924, added DNM label which i will remove one #10924 is merged.

**Comment by nordic-krch** [comment]
sample net/echo_server with netusb with log and shell enable is near its ROM limits (~99% of ROM used) on quark_se_c1000 so any increase is causing compilation failure. I already hit that in #10794 and now here. Here i solved it by disabling built-in commands in shell to save some bytes (~1k). On this branch it is now at 98,05% but this approach will not fly for long. @jukkar what would you suggest for long term?

**Comment by jukkar** [comment]
One easy way to lower the memory consumption is to lower network buffer count, but there is also a limit there after which the device is not very usable from networking point of view. What this limit is depends lot of use cases. 
@finikorg has been tweaking the netusb stuff lately with this board, do you have any suggestions for lowering memory?

File: subsys/shell/shell.c [hunk 1]
```diff
@@ -1163,6 +1163,13 @@ static void shell_log_process(const struct shell *shell)
 		processed = shell_log_backend_process(shell->log_backend);
```
Comment (pfalcon) [inline subsys/shell/shell.c]: Commit message:

> Shell log backend was enabled to early

"too early"

File: subsys/shell/Kconfig.backends [hunk 2]
```diff
@@ -18,9 +18,39 @@ config SHELL_BACKEND_SERIAL
...
+	int "Set TX ring buffer size"
+	default 16
+	depends on UART_INTERRUPT_DRIVEN
+	help
+	  If UART is utilizing DMA transfers then increasing ring buffer size
+	  increases transfers length and reduces number of interrupts.
+
+config SHELL_BACKEND_SERIAL_RX_RING_BUFFER_SIZE
+	int "Set RX ring buffer size"
+	default 8
```
Comment (pfalcon) [inline subsys/shell/Kconfig.backends]: That looks too low for me. I still didn't try to do the usual smoke test of pasting a decent length of line into the new shell - afraid of unpleasant surprises ;-). But with such a receive buffer size surprises are guaranteed. I'd suggest to set to 64 or something. On the contrary, tx buffer can decreased (if really needed.

**Comment by pfalcon** [comment]
So, as usual, this just duplicates effort of subsys/console, instead of reusing and improving it. Well, let there be parallel pathways, until they intersect. I'll be happy to give +1 when same is done to my piece, https://github.com/zephyrproject-rtos/zephyr/pull/10765 ;-).

And please specify which boards were tested with this patch - we can expect it's not working with some (or more). (That's also the reason why I don't want to make too much changes at once in #10765 - currently refactor leaves alone (mostly) known-good ISR handling).

File: samples/net/echo_server/overlay-netusb.conf [hunk 3]
```diff
@@ -7,3 +7,6 @@ CONFIG_USB_DEVICE_LOG_LEVEL_INF=y
 
 # Select USB Configurations
 CONFIG_USB_DEVICE_NETWORK_ECM=y
+
+# Disable built-in shell commands to reduce ROM footprint
+CONFIG_SHELL_CMDS=n
```
Comment (jarz-nordic) [inline samples/net/echo_server/overlay-netusb.conf]: Isn't it enough to switch off only `resize` command?

File: subsys/shell/shell.c [hunk 4]
```diff
@@ -121,17 +134,7 @@ static void shell_write(const struct shell *shell, const void *data,
 		length -= tmp_cnt;
 		if (tmp_cnt == 0 &&
 		    (shell->ctx->state != SHELL_STATE_PANIC_MODE_ACTIVE)) {
-			/* todo  semaphore pend*/
```
Comment (jarz-nordic) [inline subsys/shell/shell.c]: what happened with `todo`?

File: subsys/shell/shell.c [hunk 5]
```diff
@@ -1231,6 +1241,11 @@ void shell_thread(void *shell_handle, void *dummy1, void *dummy2)
 		return;
 	}
 
+	if ((bool)log_backend && IS_ENABLED(CONFIG_LOG)) {
+		shell_log_backend_enable(shell->log_backend, (void *)shell,
+					 (bool)log_level);
```
Comment (jarz-nordic) [inline subsys/shell/shell.c]: I do not understand this conversion. `log_level` is a pointer, you are converting it to `bool` and passing to function expecting `u32_t` ?

File: subsys/shell/shell_uart.c [hunk 6]
```diff
@@ -7,37 +7,125 @@
...
+	u8_t *data;
+	int err;
+	struct device *dev = sh_uart->ctrl_blk->dev;
+
+	len = ring_buf_get_claim(sh_uart->tx_ringbuf, &data,
+				 sh_uart->tx_ringbuf->size);
+	if (len) {
+		len = uart_fifo_fill(dev, data, len);
+		err = ring_buf_get_finish(sh_uart->tx_ringbuf, len);
+		assert(err == 0);
```
Comment (jarz-nordic) [inline subsys/shell/shell_uart.c]: 1. We shall use `__ASSERT_NO_MSG`
2. Don't you get any compiler warning about unused variable `err` when assert is off?

**Review by jakub-uC — CHANGES_REQUESTED** [review]
Please clarify some questions.

**Comment by nordic-krch** [comment]
@pfalcon 
>So, as usual, this just duplicates effort of subsys/console, instead of reusing and improving it. Well, let there be parallel pathways, until they intersect. I'll be happy to give +1 when same is done to my piece, #10765 ;-).

Yes, it was started weeks before #10765 was public. I expect they will intersect at some point. I accept you offer: you got +1 for #10765, i'm waiting for mine :).

> And please specify which boards were tested with this patch - we can expect it's not working with some (or more). (That's also the reason why I don't want to make too much changes at once in #10765 - currently refactor leaves alone (mostly) known-good ISR handling).

Tested on nrf52. ISR UART backend is use by default if platform supports uart with interrupts. Will see how many uarts implements interrupts correctly.

**Comment by finikorg** [comment]
> One easy way to lower the memory consumption is to lower network buffer count, but there is also a limit there after which the device is not very usable from networking point of view. What this limit is depends lot of use cases.
> @finikorg has been tweaking the netusb stuff lately with this board, do you have any suggestions for lowering memory?

I am afraid I took too rough approach disabling IPv6 ;)

**Review by pfalcon — APPROVED** [review]
> Yes, it was started weeks before #10765 was public. I expect they will intersect at some point. 

Yep, let's look at the bright side of it: because we follow different by close footpaths it'll allow to find the best route to place a highway, likewise, exploring different by close code will allow to explore problem space in more detail and know problem spots/find ways to address them quicker.

UART handling looks OK, approving from that side.

File: subsys/shell/Kconfig.backends (hunk 2 above)
Comment (pfalcon) [reply subsys/shell/Kconfig.backends]: > On the other hand if user is only typing then 8 is more then enough and 56 bytes of ram on some platforms means another feature.

Well, we need to think about UX and common usecases. But granted, that can be seen as requiring separate consideration (and testing), so can be done later.

**Comment by pfalcon** [comment]
@nordic-krch:

>> And please specify which boards were tested with this patch

> Tested on nrf52. ISR UART backend is use by default if platform supports uart with interrupts.

For references, subsys/console when was written was tested at least on arduino_101 (that's a "different" beast), frdm_k64f, 96b_carbon (these are about the same). Testing included being able to paste more than the rx buffer. That was one of the prompting point and requirement for subsys/console - it was developed for MicroPython porting, and it has "paste mode", which allows you to copy and paste code from a host file, manual, etc. That's the most practical way small-scale repeated interaction, and before subsys/console, nothing in Zephyr allowed pastes to work reliably.

> Will see how many uarts implements interrupts correctly.

We now have report (https://github.com/zephyrproject-rtos/zephyr/issues/10726) that subsys/console glitches on a particular STM32 chip. And as subsys/console is just a generic produces/consumer queues + ISR, then the problem would be either in UART driver, or that this particular chip has IRQ process not fitting into the model (and I spent enough time to chart what would be the appropriate IRQ handling model for UART drivers, again, it work for as diverse cases as 16550 (arduino_101) and common single-byte RX/TX regs as a baselines for ARM chips). My current mental solution for that is console/subsys (to be upgraded to ring_buffer, like here) is a *reference*, baseline implementation. It should move to drivers/serial, and a particular driver can either use it, or can override completely. All that are just draft thoughts, should be considered for interaction with https://github.com/zephyrproject-rtos/zephyr/pull/10820, etc.

File: subsys/shell/shell.c (hunk 4 above)
Comment (nordic-krch) [reply subsys/shell/shell.c]: `todo` was outdated and should have been removed when `k_poll` was added

File: subsys/shell/shell.c (hunk 5 above)
Comment (nordic-krch) [reply subsys/shell/shell.c]: fixed. I also refactored a bit to cast early those `void *` to local variables.

**Review by jakub-uC — APPROVED** [review]

**Comment by nordic-krch** [comment]
@jarz-nordic please take another look, i've added blocking support to the backend (to handle panic case). Also moved rx polling period to kconfig.

File: subsys/shell/Kconfig.backends [hunk 7]
```diff
@@ -18,9 +18,46 @@ config SHELL_BACKEND_SERIAL
 	bool "Enable serial backends."
 	default y
 	select SERIAL
+	select RING_BUFFER
 	help
 	  Enable serial backends.
 
+# Internal config to enable UART interrupts if supported.
+config SHELL_BACKEND_SERIAL_FORCE_INTERRUPTS
```
Comment (jarz-nordic) [inline subsys/shell/Kconfig.backends]: Maybe rename it to `SHELL_BACKEND_SERIAL_USE_INTERRUPTS`

**Review by jakub-uC — APPROVED** [review]

**Review by nashif — APPROVED** [review]

**Comment by pfalcon** [comment]
@nordic-krch: So, how well did you test this patch with various qemu targets? Are you aware of https://github.com/zephyrproject-rtos/zephyr/issues/8869 for example? In summary, interrupt-driven UART is great boon for real hardware, and curse for QEMU emulations, which don't have faithful IRQ emulation, sometimes even remotely.

All in all, I see obvious shell behavior regressions with this patch for BOARD=mps2_an385 in qemu (and that's about to become the default qemu target for arm).

**Comment by jarz-nordic** [comment]
@pfalcon : Old shell was also interrupt driven for Rx.

**Comment by pfalcon** [comment]
> @pfalcon : Old shell was also interrupt driven for Rx.

Old shell as in "old shell", or as in "new shell before this PR"? The old shell had a pretty fun design, which didn't work well even on real hardware (e.g. for pastes). Here we discuss the problem with QEMU and this particular PR.

**Comment by jarz-nordic** [comment]
@pfalcon : old shell as a old shell. So this is not a regression but something we need to improve :)

File: subsys/shell/Kconfig.backends [hunk 8]
```diff
@@ -18,9 +18,46 @@ config SHELL_BACKEND_SERIAL
...
+	bool
+	default y
+	depends on SERIAL_SUPPORT_INTERRUPT
+	imply UART_INTERRUPT_DRIVEN
```
[the remark this answers is not in the document]
Comment (jarz-nordic) [reply subsys/shell/Kconfig.backends]: It is nice difference. With `imply` you are able to unselect this option with menuconfig.

**Comment by pfalcon** [comment]
@jarz-nordic :

> old shell as an old shell. So this is not a regression but something we need to improve :)

Yeah, we mostly left old shell behind (https://github.com/zephyrproject-rtos/zephyr/issues/10769 is again mostly to prove the point that if we replace "old" with "new", we should port over the functionality *by default*, or someone will ask why we didn't, and we'll need to have an answer, preferably good).

So, this is not regression re: old shell. This is qemu-specific regression before and after this patch. You're right that we need to deal with it. Let's move discussion to a more suitable place, I'll see if we can repurpose some existing ticket or it's better to create new.

**Comment by nordic-krch** [comment]
@pfalcon 
> So, how well did you test this patch with various qemu targets? 

Well, i think that we have uart.h api that i should be able to rely on. If specific driver implements interrupt version then it's not my fault if it is faulty :). If qemu interrupt driven uart is lousy then it should be improved or disabled or we should have big banner somewhere `do not trust drivers, test all before any change`

**Comment by pfalcon** [comment]
@nordic-krch, @jarz-nordic: Let's move discussion to https://github.com/zephyrproject-rtos/zephyr/issues/8869 please.

**Pull request merged**